File: sandbox/backend/java/simple-eccomerce/src/main/java/com/simple/ecommerce/model/Product.java
package com.simple.ecommerce.model;

/**
 * Represents a product in the eCommerce system.
 * Business Rules:
 * - Must have name, price and category
 * - Can be marked available or unavailable
 */
public class Product {
    private String name;
    private double price;
    private Category category;
    private boolean available;

    /**
     * Creates a new product with the specified attributes.
     *
     * @param name     the product name
     * @param price    the product price
     * @param category the product category
     */
    public Product(String name, double price, Category category) {
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("Product name cannot be empty");
        }
        if (price < 0) {
            throw new IllegalArgumentException("Product price cannot be negative");
        }
        if (category == null) {
            throw new IllegalArgumentException("Product category cannot be null");
        }

        this.name = name;
        this.price = price;
        this.category = category;
        this.available = true; // By default, products are available
    }

    /**
     * Returns the product name.
     *
     * @return the product name
     */
    public String getName() {
        return name;
    }

    /**
     * Returns the product price.
     *
     * @return the product price
     */
    public double getPrice() {
        return price;
    }

    /**
     * Returns the product category.
     *
     * @return the product category
     */
    public Category getCategory() {
        return category;
    }

    /**
     * Checks if the product is available.
     *
     * @return true if the product is available, false otherwise
     */
    public boolean isAvailable() {
        return available;
    }

    /**
     * Sets the product availability status.
     *
     * @param available the availability status to set
     */
    public void setAvailable(boolean available) {
        this.available = available;
    }

    @Override
    public String toString() {
        return name + " ($" + price + ")";
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Product product = (Product) obj;
        return name.equals(product.name);
    }

    @Override
    public int hashCode() {
        return name.hashCode();
    }
} 

File: Projects/Testtrack/p911 - Simple eCommerce backend/domain/entity/product.py
from typing import Self
from domain.entity.category import Category


class Product:
    def __init__(self, name: str, price: float, category: Category, available: bool = True):
        if name == None or name.strip() == None:
            raise ValueError("Product name cannot be empty")
        else:
            self.name = name

        if price < 0:
            raise ValueError("product price cannot be negative")
        else:
            self.price = price

        if category == None:
            raise ValueError("Progduct category cannot be empty")
        else:
            self.category = category

        self.available = available

    def __eq__(self, other) -> bool:
        if self is other:
            return True
        if not isinstance(other, Product):
            return False
        return self.name == other.name

    def __hash__(self) -> int:
        return hash(self.name)

    def __repr__(self) -> str:
        return f"Product(name='{self.name}')"

    def __str__(self):
        return f"{self.name} ${self.price}"

    def get_name(self) -> str:
        return self.name

    def get_price(self) -> str:
        return self.price

    def get_category(self) -> Category:
        return self.category

    def is_available(self) -> bool:
        return self.available

    def set_available(self, available: bool) -> Self:
        self.available = available

        return self
